### DISCUSSION TITLE
What is an NAT tun?

### DISCUSSION THREAD

**Discussion by klzgrad** [SagerNet/SagerNet#327] [Q&A] [answered]
https://github.com/SagerNet/SagerNet/commit/416c8e491c362951806af8acd4bdae384a02a610

I thought configuring NAT requires root. How does it work?

**Comment by nekohasekai** [comment, marked as answer]
It's just the same principle as the previous native forwarding, but with golang and gVisor stack implementation , with no socks5.

It was made because it was always said that Clash for Android was faster, and CFA uses a [similar implementation](https://github.com/Kr328/tun2socket) (mostly derived from clash-premium's system stack tun).

Now we have the same thing, including IPv6 support. But performance varies from device to device, some people have no difference, others feedback that system tun is much faster in single threaded connections.

This can also be used as a replacement in some devices where the gVisor TCP connection fails for unknown reasons (mostly happens in low visibility custom AOSP, almost never in stock ROM).

**Reply by klzgrad** [reply]
It's with gVisor stack, but it is also used as a replacement for gVisor? Which is it then?

The /dev/tun device reads and writes in packets. Lwip reassembles the packets into TCP streams. The streams are then hooked up to higher protocol stacks. Where does the "NAT tun" come into play here? Who is doing the "NAT", the Android Linux kernel or gVisor?

**Reply by nekohasekai** [reply]
> It's with gVisor stack, but it is also used as a replacement for gVisor? Which is it then?

It does not reassemble packets, but rewrites TCP/IP headers.

> Where does the "NAT tun" come into play here?

It uses gVisor to parse and rewrite packets so that TCP connections are redirected to a local TCP server, and then gets the destination address based on the mapped port.

origin: 

`source address:source port -> destination address:destination port`

rewritten: 

`destination address:source port -> local TCP server`

If the packet originates from the local TCP server, the address is restored and written out.

For UDP packets, the implementation is a bit simpler, but this works just as well.

> Who is doing the "NAT", the Android Linux kernel or gVisor?

Probably the tools provided by gVisor, although it is not difficult to re-implement.

**Reply by klzgrad** [reply]
Is it correct to read this as a TPROXY in gVisor? It removes the overhead of socks5.

**Reply by nekohasekai** [reply]
Yes, but the point is not socks5.
